src/platform_impl/linux/x11/ime/context.rs

```diff
@@ -1,41 +1,239 @@
 use std::{
+    mem::{transmute, MaybeUninit},
     os::raw::{c_short, c_void},
     ptr,
     sync::Arc,
 };
 
 use super::{ffi, util, XConnection, XError};
+use crate::event::CompositionEvent::{CompositionEnd, CompositionStart, CompositionUpdate};
+use crate::platform_impl::platform::x11::ime::ImeEventSender;
+use std::ffi::CStr;
+use std::iter::FromIterator;
+use x11_dl::xlib::{XIMPreeditCaretCallbackStruct, XIMPreeditDrawCallbackStruct};
 
 #[derive(Debug)]
 pub enum ImeContextCreationError {
     XError(XError),
     Null,
 }
+type XIMProcNonnull = unsafe extern "C" fn(ffi::XIM, ffi::XPointer, ffi::XPointer);
+extern "C" fn preedit_start_callback(
+    _xim: ffi::XIM,
+    client_data: ffi::XPointer,
+    _call_data: ffi::XPointer,
+) -> i32 {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+
+    client_data.text.clear();
+    client_data.cursor_pos = 0;
+    client_data.is_composing = true;
+    client_data
+        .event_sender
+        .send((client_data.window, CompositionStart()))
+        .expect("failed to send composition start event");
+    debug!("preedit start callback");
+    -1
+}
+
+extern "C" fn preedit_done_callback(
+    _xim: ffi::XIM,
+    client_data: ffi::XPointer,
+    _call_data: ffi::XPointer,
+) {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+
+    if !client_data.text.is_empty() {
+        client_data.text.clear();
+        client_data.cursor_pos = 0;
+        client_data
+            .event_sender
+            .send((
+                client_data.window,
+                CompositionUpdate(
+                    String::from_iter(client_data.text.clone()),
+                    client_data.cursor_pos,
+                ),
+            ))
+            .expect("failed to send composition update event");
```

Comment:
This code generates a `CompositionUpdate` before each `CompositionEnd`, always with an empty buffer. I don't believe this is desired behavior. (No such event is generated by Firefox [in this demo](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event#Result).)

Reply:
I think it depends on the IME behaviour. At least, in my environment, when I hit enter to decide the preedit string, firstly, preedit_update_callback is called to delete all preedit string, and then preedit_done_callback is called. In this case, CompositionUpdate won't be generated here.
This might be a little bit different from Firefox demo. Firefox itself uses GTK to manage those IME events, and I referred to the logic in GTK. So, this behaviour is similar to GTK. It seems Firefox fileter those event to align Web specifications.
I'm not sure what is the best way. I though that it was better to expose those events as much as possible, and each application using winit should filter them as they need. But maybe it is better to follow Web specification to provide unified interfaces.

Do you think those filtering should happen inside the library?

Reply:
I checked the [UI events spec](https://w3c.github.io/uievents/#event-type-compositionupdate) for clarification on this and my understanding is that a `CompositionUpdate` event with an empty buffer value is generated when the user has cancelled the IME input. So, I think that, if the IME has actually been completed, a `CompositionUpdate("", 0)` event should not be generated before `CompositionEnd`. In my testing, I am receiving an empty update event immediately before the end event, after either cancelling or completing input.

Reply:
Changed. I added new states in event_processor and convert original events to events following this event s spec.